### PR TITLE
[WIP] placement scheduling based on cluster resource usage

### PR DESCRIPTION
**Pull request by haoqing0110**
ref: https://github.com/open-cluster-management/backlog/issues/14755
Signed-off-by: haoqing0110 <qhao@redhat.com>

### CODE REVIEW

File: pkg/controllers/scheduling/schedule.go [hunk 1]
```diff
@@ -140,6 +142,12 @@ func (s *pluginScheduler) Schedule(
 	for _, cluster := range filtered {
 		scoreSum[cluster.Name] = 0
 	}
+	// define steady prioritizer weight, default is PolicyTypeNotSteady
+	steadyWeight := 0.1
+	if placement.Spec.ChurningPolicy.PolicyType == clusterapiv1alpha1.PolicyTypeSteady {
+		steadyWeight = float64(len(s.prioritizers) + 1)
```
Comment (qiujian16) [inline pkg/controllers/scheduling/schedule.go:148]: I think you would want a map[string]int for weight and give it init values. You could have a separate func to adjust weight.

File: pkg/plugins/resource/resource.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			}
+		}
+
+		maxScore = max(maxScore, scores[cluster.Name])
+	}
+
+	// normalize the score and ensure the value falls in the range between 0 and 100.
+	for _, cluster := range clusters {
+		if maxScore != 0 {
+			scores[cluster.Name] = scores[cluster.Name] * 100 / maxScore
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:63]: make it range from -100  to 100.

File: pkg/plugins/resource/resource.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+	for _, cluster := range clusters {
+		acpu := cluster.Status.Allocatable[clusterapiv1.ResourceCPU]
+		ccpu := cluster.Status.Capacity[clusterapiv1.ResourceCPU]
+		amem := cluster.Status.Allocatable[clusterapiv1.ResourceMemory]
+		cmem := cluster.Status.Capacity[clusterapiv1.ResourceMemory]
+
+		for _, v := range placement.Spec.ResourceUsagePreferences {
+			if v.ResourceName == clusterapiv1alpha1.ResourceNameCPU {
+				scores[cluster.Name] += int64(acpu.AsApproximateFloat64() * 100 / ccpu.AsApproximateFloat64())
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:50]: I think you would need to handle the case that capacity is 0.

File: pkg/plugins/resource/resource.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	scores := map[string]int64{}
+	var maxScore int64
+
+	for _, cluster := range clusters {
+		acpu := cluster.Status.Allocatable[clusterapiv1.ResourceCPU]
+		ccpu := cluster.Status.Capacity[clusterapiv1.ResourceCPU]
+		amem := cluster.Status.Allocatable[clusterapiv1.ResourceMemory]
+		cmem := cluster.Status.Capacity[clusterapiv1.ResourceMemory]
+
+		for _, v := range placement.Spec.ResourceUsagePreferences {
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:48]: make it a separated  func, so you could reduce  the  nested level

File: pkg/controllers/scheduling/schedule.go [hunk 5]
```diff
@@ -140,6 +142,12 @@ func (s *pluginScheduler) Schedule(
 	for _, cluster := range filtered {
 		scoreSum[cluster.Name] = 0
 	}
+	// define steady prioritizer weight, default is PolicyTypeNotSteady
+	steadyWeight := 0.1
+	if placement.Spec.ChurningPolicy.PolicyType == clusterapiv1alpha1.PolicyTypeSteady {
```
Comment (dongwangdw) [inline pkg/controllers/scheduling/schedule.go:147]: looks like ChurningPolicy is optional, shall we check its value here?

Comment (haoqing0110) [reply pkg/controllers/scheduling/schedule.go:147]: ChurningPolicy is out of scope based on latest https://github.com/open-cluster-management-io/enhancements/pull/16/files, will remove this part.

File: pkg/plugins/resource/resource.go (hunk 3 above)
Comment (haoqing0110) [reply pkg/plugins/resource/resource.go]: fixed.

File: pkg/plugins/resource/resource.go (hunk 2 above)
Comment (haoqing0110) [reply pkg/plugins/resource/resource.go]: As discussed, will range it from 0 - 100.

File: pkg/controllers/scheduling/schedule.go (hunk 1 above)
Comment (haoqing0110) [reply pkg/controllers/scheduling/schedule.go]: As discussed, remove steady related code.

**Comment by haoqing0110** [comment]
Close as a new PR created in https://github.com/open-cluster-management-io/placement/pull/32.

**Pull request closed without merging**